README
Fast Auxiliary Space Preconditioning (FASP) Solver Library: README

Last modified on 05/08/2021

================================================================================

This software is free software distributed under the Lesser General Public 
License or LGPL, version 3.0 or any later versions. This software distributed 
in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even 
the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
See the GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License 
along with FASP. If not, see <http://www.gnu.org/licenses/>.

The FASP package is designed for developing and testing new efficient solvers 
and preconditioners for discrete partial differential equations (PDEs) or 
systems of PDEs. The main components of the package are standard Krylov methods, 
algebraic multigrid methods, and incomplete factorization methods. Based on 
these standard techniques, we build efficient solvers, based on the framework 
of Auxiliary Space Preconditioning, for several complicated applications. 
Current examples include the fluid dynamics, underground water simulation, 
the black oil model in reservoir simulation, and so on. 

To compile, you need a C99 compiler and a F90 compiler. By default, we use GNU
gcc/gfortan, respectively; see INSTALL for more details.

Configuring and building the FASP library and test suite requires CMake 2.6 or
higher, which can be found at http://www.cmake.org/.

The command to configure is:

    $ mkdir Build; cd Build; cmake .. 

After successfully configing the environment, just run:

    $ make                   // to compile the FASP static or shared library
    
To install the FASP library and executables, run:

    $ make install
	
Note: The default prefix is the FASP source directory.

================================================================================

This package has been tested with on the following platforms: 

--Linux: GNU gcc/gfortran, Intel icc/ifort

--Mac OS X: GNU gcc/gfortran, Intel icc/ifort, Apple clang
    
--Windows XP, 7, 10: Intel icc/ifort

================================================================================
    
More details about the FASP project can be found at http://fasp.sourceforge.net.
And, if you have any questions, feel free to contact us at faspdev@gmail.com.


log/CHANGES
File Name: CHANGES

02/11/2022 FASP 2.8.6:
- Add weighted Jacboi for F-points;
- Add reduction-based AMG;
- Test reduction-based AMG;
- Remove cxx_fc_compilers;
- Update cmake for WITHFC;
- Fix rpath problem in MacOS;
- Update VS22 solution for Strumpack.

10/16/2022 FASP 2.8.5:
- Use clang-format from OCP;
- Add GSF smoother (GS for F-points);
- Add a Fortran wrapper for Strumpack.

10/09/2022 FASP 2.8.4:
- Remove xcode support, move to VSCode+cmake completely;
- Add and reset .vscode settings;
- Update git ignores;
- Simplify cmake files;
- Test on arm64 computer.

10/03/2022 FASP 2.8.3:
- Add STRUMPACK as a direct solver;
- Wrong solution when used with STRUMPACK_METIS re-ordering for old METIS;
- Updated to METIS-5.1.0; seems to work using STRUMPACK_METIS now;;
- Update FindMETIS.cmake.

09/26/2022 FASP 2.8.2:
- Chang default behavoir: Do not require Fortran compilers!!!
- Do not compile Fortran by default;
- Use -DWITHFC=ON in cmake to enable Fortran support.

09/21/2022 FASP 2.8.1:
- Add OpenMP DLL version for Windows;
- Longer output format for residuals.

09/06/2022 FASP 2.8.0:
- Major Change: Need to update all FASP projects;
- Add abstol to KSM functions;
- Prepare for Iterative Refinement;
- Add LONGREAL (long double by default).

09/03/2022 FASP 2.7.9:
- Unify number of digits for output.

08/30/2022 FASP 2.7.8:
- Output stat information from MUMPS;
- Output stat information from UMFPACK;
- Output stat information from PARDISO;
- Test MUMPS, UMFPACK, and PARDISO with problems from DLD and Huawei;
- MUMPS-5.5.1 built with gfortran (does not work with ifort);
- SuiteSparse-5.12.0 built with clang (work with icc/gcc);
- Remove the GUI installation tcl/tk script;
- Update the outdated User Guide!!!

08/28/2022 FASP 2.7.7:
- Add output subroutine for MTX format;
- Test PARDISO and MUMPS;
- Modify the main test example: test.c;
- Fix a coarse level dof size bug: SHORT --> INT;
- Report MUMPS seg fault bug on OSX;
- Update documantation and webpage;
- Fix Doxygen problems.

08/28/2022 FASP 2.7.6:
- Prepare for Huawei's test problems;
- Merge different versions;
- Add MKL support for Mac OS X;
- Add METIS and GKlib support;
- Test linking to MUMPS;
- Test linking to MKL Pardiso;
- Add reading subroutine for Huawei's problems;
- Add support to MUMPS and Pardiso; need to test;
- Add Huawei test problems;
- Add DLD test problems.

08/14/2022 FASP 2.7.5:
- Checking return status of UMFPACK;
- Fixing regression test failure of fasp4blkoil.

07/27/2022 FASP 2.7.4:
- Fix Pardiso and Umfpack timer output;
- Test OpenMP for DLD problem.

04/08/2022 FASP 2.7.3:
- Reading input parameters from a file for fasp4ns project;
- Remove unused variables in UMFPACK interface;
- Update Fortran wrapper for Krylov_AMG;

03/08/2022 FASP 2.7.2:
- Fix a bug due the ISNAN macro not recognized by some compilers;
- Use isnan() to replace ISNAN;
- Fix a TeX problem with Doxygen;
- Fix math.lib linkage problem;
- Fix cmake debug mode for Windows;
- Add oneAPI MKL support for Linux;
- Update webpage;
- Update Doxygen.

01/27/2022 FASP 2.7.1:
- Add VS2022 + Intel 2022 oneAPI solution file;
- Fix unused variables and other warnings by oneAPI;
- Update webpage;
- Update Doxygen.

08/14/2021 FASP 2.7.0:
- Update webpage;
- Update warning level in UA AMG;
- Update Doxygen;
- Git becomes public.

05/12/2021 FASP 2.6.8:
- Update README and cmake examples;
- Remove support for VS10 and VS15;
- Update webpage and documentation;
- Update VS19 setting with more aggressive optimization;
- Change README to MarkDown format;
- Update matrix inversion subroutine and give return values.

04/27/2021 FASP 2.6.7:
- Eamon Zhang reported several bugs;
- Fix the bug caused by inversion of smat matrices;
- Fix the bug cuased by uninitialized ILU param;
- Fix other small problems in DEBUG mode;
- Fix a bug in external.cmake for linking to MKL etc.

03/23/2021 FASP 2.6.6:
- Update CMake and add backup target;
- Saving coarse matrices for debugging.

03/13/2021 FASP 2.6.5:
- Fix DEBUG information in SmatInv;
- Fix Windows VS19 OpenMP flag;
- Test OpenMP for icc;
- Rewrite cmake: Replace FASP_SOURCE_DIR with general dir.

03/05/2021 FASP 2.6.4:
- Remove testlib from VS19;
- Add DLL version for Windows;
- Fix DLL version for vscode/cmake/msvc;
- Fix OpenMP version for cmake.

02/13/2021 FASP 2.6.3:
- Remove BLAS from CMake;
- Make CMake work on Windows + VSCode + CMake;
- Fix MSVC flag: _CRT_SECURE_NO_WARNINGS.

02/10/2021 FASP 2.6.2:
- Update output information for UA AMG setup;
- Add decoup_type for choosing decoupling methods in BSR;
- Update test problems in main/test.c;
- Update CMake external packages.

02/05/2021 FASP 2.6.1:
- Add uninstall target to the new CMakeFile.txt;
- Now both old and new cmake work!
- Change output of fasp_solver_amg();
- Update header file;
- Update return values for AMG solvers;
- Add Release/Debug targets into cmake.

01/25/2021 FASP 2.6.0:
- Switch cmake style.
- Use the same compile and link style as fasp4blkoil.
- Check compatibility with PennSim new version!
- Update benchmark test input format.
- Double-check PARDISO interface.

01/15/2021 FASP 2.5.9:
- Fix divide-by-zero in GCG for NAMLI.
- Fix some typos in comments.

01/10/2021 FASP 2.5.8:
- Add benchmark test based on SuiteSparse Matrix Market.
- Modify test/main/test.c slightly.
- Update error messages for reading files.
- Rewrite fasp_dbsr_print.
- Make a separate dir for benchmark test.
- Fix problems with icc compiler for Linux.

12/15/2020 FASP 2.5.7:
- Check trivial solution case Ax=0 for AMG cycles.
- Update cmake required version.
- Update cmake CC search list.
- Do not check for Doxygen by default.
- Add VW and WV hybrid cycles.

11/04/2020 FASP 2.5.6:
- Fix new warnings in GCC10.
- Fix unsigned long problem for INT.
- Fix debug mode.
- Update version information.

10/13/2020 FASP 2.5.5:
- Important fixes to UA AMG: handle possible divide by zero error.
- Update INT definition and test INT=int and long.

base/src/vec.c
/*! \file vec.c
 *
 *  \brief Simple operations for vectors
 *
 *  \note All structures should be initialized before usage.
 */

#include <math.h>

#ifdef _OPENMP
#include <omp.h>
#endif

#include "fasp.h"
#include "fasp_functs.h"

/*---------------------------------*/
/*--      Public Functions       --*/
/*---------------------------------*/

/**
 * \fn INT fasp_dvec_isnan (dvector *u)
 *
 * \brief Check a dvector whether there is NAN
 *
 * \param u    Pointer to dvector
 *
 * \return     Return TRUE if there is NAN
 *
 * \author Chensong Zhang
 * \date   2013/03/31
 */
INT fasp_dvec_isnan (dvector *u)
{
    INT i;
    
    for ( i=0; i<u->row; i++ ) {
        if ( ISNAN(u->val[i]) ) return TRUE;
    }

    return FALSE;
}

/**
 * \fn dvector fasp_dvec_create (const INT m)
 *
 * \brief Create dvector data space of REAL type
 *
 * \param m    Number of rows
 *
 * \return u   The new dvector
 *
 * \author Chensong Zhang 
 * \date   2010/04/06
 */
dvector fasp_dvec_create (const INT m)
{
    dvector u;
    
    u.row = m;
    u.val = (REAL *)fasp_mem_calloc(m,sizeof(REAL));
    
    return u;
}

/**
 * \fn ivector fasp_ivec_create (const INT m)
 *
 * \brief Create vector data space of INT type
 *
 * \param m   Number of rows
 *
 * \return u  The new ivector
 *
 * \author Chensong Zhang 
 * \date   2010/04/06
 */
ivector fasp_ivec_create (const INT m)
{    
    ivector u;
    
    u.row = m;
    u.val = (INT *)fasp_mem_calloc(m,sizeof(INT)); 
    
    return u;
}

/**
 * \fn void fasp_dvec_alloc (const INT m, dvector *u)
 *
 * \brief Create dvector data space of REAL type
 *
 * \param m    Number of rows
 * \param u    Pointer to dvector (OUTPUT)
 *
 * \author Chensong Zhang 
 * \date   2010/04/06
 */
void fasp_dvec_alloc (const INT  m,
                      dvector   *u)
{    
    u->row = m;
    u->val = (REAL*)fasp_mem_calloc(m,sizeof(REAL)); 
    
    return;
}

/**
 * \fn void fasp_ivec_alloc (const INT m, ivector *u)
 *
 * \brief Create vector data space of INT type
 *
 * \param m   Number of rows
 * \param u   Pointer to ivector (OUTPUT)
 *
 * \author Chensong Zhang 
 * \date   2010/04/06
 */
void fasp_ivec_alloc (const INT  m,
                      ivector   *u)
{    
    
    u->row = m;
    u->val = (INT*)fasp_mem_calloc(m,sizeof(INT));
    
    return;
}

/**
 * \fn void fasp_dvec_free (dvector *u)
 *
 * \brief Free vector data space of REAL type
 *
 * \param u   Pointer to dvector which needs to be deallocated
 *
 * \author Chensong Zhang
 * \date   2010/04/03  
 */
void fasp_dvec_free (dvector *u)
{    
    if (u==NULL) return;
    
    fasp_mem_free(u->val);
    u->row = 0; u->val = NULL; 
}

/**
 * \fn void fasp_ivec_free (ivector *u)
 *
 * \brief Free vector data space of INT type
 *
 * \param u   Pointer to ivector which needs to be deallocated
 *
 * \author Chensong Zhang
 * \date   2010/04/03  
 *
 * \note This function is same as fasp_dvec_free except input type.
 */
void fasp_ivec_free (ivector *u)
{    
    if (u==NULL) return;
    
    fasp_mem_free(u->val);
    u->row = 0; u->val = NULL; 
}

/**
 * \fn void fasp_dvec_null (dvector *x) 
 *
 * \brief Initialize dvector
 *
 * \param x   Pointer to dvector which needs to be initialized
 *
 * \author Chensong Zhang
 * \date 2010/04/03  
 */
void fasp_dvec_null (dvector *x) 
{
    x->row = 0; x->val = NULL;
}

/**
 * \fn void fasp_dvec_rand (const INT n, dvector *x)
 *
 * \brief Generate random REAL vector in the range from 0 to 1
 *
 * \param n    Size of the vector
 * \param x    Pointer to dvector
 * 
 * \note Sample usage: 
 * \par
 *   dvector xapp;
 * \par
 *   fasp_dvec_create(100,&xapp);
 * \par
 *   fasp_dvec_rand(100,&xapp);
 * \par
 *   fasp_dvec_print(100,&xapp);
 *
 * \author Chensong Zhang
 * \date   11/16/2009
 */
void fasp_dvec_rand (const INT  n,
                     dvector   *x)
{
    const INT va=(REAL) 0;
    const INT vb=(REAL) n;
    
    INT s=1; srand(s);
    
    INT i,j;
    
    x->row = n;
    
    for (i=0; i<n; ++i){
        j = 1 + (INT) (((REAL)n)*rand()/(RAND_MAX+1.0));
        x->val[i] = (((REAL)j)-va)/(vb-va);
    }
}

/**
 * \fn void fasp_dvec_set (INT n, dvector *x, REAL val)
 *
 * \brief Initialize dvector x[i]=val for i=0:n-1
 *
 * \param n      Number of variables
 * \param x      Pointer to dvector
 * \param val    Initial value for the vector
 *
 * \author Chensong Zhang
 * \date   11/16/2009
 *
 * Modified by Chunsheng Feng, Xiaoqiang Yue on 05/23/2012    
 */
void fasp_dvec_set (INT      n,
                    dvector *x, 
                    REAL     val)
{
    INT i;
    REAL *xpt=x->val;
    
    if (n>0) x->row=n; 
    else n=x->row; 
	
#ifdef _OPENMP 
    // variables for OpenMP 
    INT myid, mybegin, myend;
    INT nthreads = FASP_GET_NUM_THREADS();
#endif
   
    if (val == 0.0) {
        
#ifdef _OPENMP 
        if (n > OPENMP_HOLDS) {
#pragma omp parallel for private(myid, mybegin, myend) 
            for (myid = 0; myid < nthreads; myid++ ) {
                FASP_GET_START_END(myid, nthreads, n, &mybegin, &myend);
                memset(&xpt[mybegin],0x0, sizeof(REAL)*(myend-mybegin));
            }
        }
        else {
#endif
            memset(xpt, 0x0, sizeof(REAL)*n);
#ifdef _OPENMP
        }
#endif
        
    }
    
    else {
        
#ifdef _OPENMP 
        if (n > OPENMP_HOLDS) {
#pragma omp parallel for private(myid, mybegin, myend) 
            for (myid = 0; myid < nthreads; myid++ ) {
                FASP_GET_START_END(myid, nthreads, n, &mybegin, &myend);
                for (i=mybegin; i<myend; ++i) xpt[i]=val;
            }
        }
        else {
#endif
            for (i=0; i<n; ++i) xpt[i]=val;
#ifdef _OPENMP
        }
#endif
        
    }
}

/**
 * \fn void fasp_ivec_set (const INT m, ivector *u)
 *
 * \brief Set ivector value to be m
 *
 * \param  m    Integer value of ivector
 * \param  u    Pointer to ivector (MODIFIED)
 *
 * \author Chensong Zhang
 * \date   04/03/2010  
 *
 * Modified by Chunsheng Feng, Xiaoqiang Yue  
 * \date   05/23/2012    
 */
void fasp_ivec_set (const INT  m,
                    ivector   *u)
{    
        INT i;
        INT n = u->row;
        INT nthreads = 1, use_openmp = FALSE;
    
#ifdef _OPENMP  
	if ( n > OPENMP_HOLDS ) {
            use_openmp = TRUE;
            nthreads = FASP_GET_NUM_THREADS();
	}
#endif
    
	if (use_openmp) {
		INT mybegin, myend, myid;
#ifdef _OPENMP 
#pragma omp parallel for private(myid, mybegin, myend, i) 
#endif
        for (myid = 0; myid < nthreads; myid++ ) {
            FASP_GET_START_END(myid, nthreads, n, &mybegin, &myend);
            for (i=mybegin; i<myend; ++i) u->val[i] = m;
        }        
	}    
	else {
        
            for (i=0; i<n; ++i) u->val[i] = m;
	}
}

/**
 * \fn void fasp_dvec_cp (dvector *x, dvector *y) 
 *
 * \brief Copy dvector x to dvector y
 *
 * \param x  Pointer to dvector
 * \param y  Pointer to dvector (MODIFIED)
 *
 * \author Chensong Zhang
 * \date   11/16/2009
 */
void fasp_dvec_cp (dvector *x, 
                   dvector *y)
{
    y->row=x->row;
    memcpy(y->val,x->val,x->row*sizeof(REAL));
}

/**
 * \fn REAL fasp_dvec_maxdiff (dvector *x, dvector *y) 
 *
 * \brief Maximal difference of two dvector x and y
 *
 * \param  x    Pointer to dvector
 * \param  y    Pointer to dvector
 *
 * \return      Maximal norm of x-y
 *
 * \author Chensong Zhang
 * \date   11/16/2009
 *
 * Modified by chunsheng Feng, Zheng Li
 * \date   06/30/2012
 */
REAL fasp_dvec_maxdiff (dvector *x, 
                        dvector *y)
{
    const INT length=x->row;
    REAL Linf=0., diffi=0.;
    REAL *xpt=x->val, *ypt=y->val;
    
    INT use_openmp = FALSE;
    INT i;

#ifdef _OPENMP 
    INT myid, mybegin, myend, nthreads;
    if ( length > OPENMP_HOLDS ) {
        use_openmp = TRUE;
        nthreads = FASP_GET_NUM_THREADS();
    }
#endif
    
    if(use_openmp) {
#ifdef _OPENMP 
        REAL temp = 0.;
#pragma omp parallel firstprivate(temp) private(myid, mybegin, myend, i, diffi) 
        {
            myid = omp_get_thread_num();
            FASP_GET_START_END(myid, nthreads, length, &mybegin, &myend);
            for(i=mybegin; i<myend; i++) {
                if ((diffi = ABS(xpt[i]-ypt[i])) > temp) temp = diffi;
            }
#pragma omp critical
            if(temp > Linf) Linf = temp;
        }
#endif
    }
    else {
        for (i=0; i<length; ++i) {
            if ((diffi = ABS(xpt[i]-ypt[i])) > Linf) Linf = diffi;
        }
    }
    
    return Linf;
}

/**
 * \fn void fasp_dvec_symdiagscale (dvector *b, dvector *diag)
 *
 * \brief Symmetric diagonal scaling D^{-1/2}b
 *
 * \param b       Pointer to dvector
 * \param diag    Pointer to dvector: the diagonal entries
 *
 * \author Xiaozhe Hu
 * \date   01/31/2011
 */
void fasp_dvec_symdiagscale (dvector *b, 
                             dvector *diag)
{
    // information about dvector
    const INT n = b->row;
    REAL *val = b->val;
    
    // local variables
    INT i;
    
    if (diag->row != n) {
        printf("### ERROR: Sizes of diag = %d != dvector = %d!", diag->row, n);
        fasp_chkerr(ERROR_MISC, __FUNCTION__);
    }
    
    INT use_openmp = FALSE;
    
#ifdef _OPENMP 
    INT mybegin, myend, myid, nthreads;
    if ( n > OPENMP_HOLDS ){
        use_openmp = TRUE;
        nthreads = FASP_GET_NUM_THREADS();
    }
#endif

    if (use_openmp) {
#ifdef _OPENMP 
#pragma omp parallel for private(myid, mybegin,myend) 
        for (myid = 0; myid < nthreads; myid++ ) {
            FASP_GET_START_END(myid, nthreads, n, &mybegin, &myend);
            for (i=mybegin; i<myend; ++i) val[i] = val[i]/sqrt(diag->val[i]);
        }        
#endif
    }    
    else {
        for (i=0; i<n; ++i) val[i] = val[i]/sqrt(diag->val[i]);
    }
    
    return;
}

/*---------------------------------*/
/*--        End of File          --*/
/*---------------------------------*/
